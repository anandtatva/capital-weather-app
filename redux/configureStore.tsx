import { createStore, combineReducers } from "redux";
import { persistReducer, persistStore } from "redux-persist"
import AsyncStorage from "@react-native-async-storage/async-storage"
import CountrySlice from "./reducer/country-slice"
const persistConfig = {
    key: "country",
    storage: AsyncStorage
}

const rootReducer = combineReducers({
    country: CountrySlice
})

const persistReducers = persistReducer(persistConfig, rootReducer)

const store = createStore(persistReducers);

const persistor = persistStore(store)

export {
    store,
    persistor
}