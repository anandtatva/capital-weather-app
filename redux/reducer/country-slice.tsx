import { createSlice } from "@reduxjs/toolkit";
import { CountryType } from "../../types/country-type"
import { CapitalWeatherType } from "../../types/capital-weather-type"


interface CountryState {
    name: string;
    countries: CountryType[];
    capitalWeather: CapitalWeatherType | null;
}

const initialState: CountryState = {
    name: "",
    capitalWeather: null,
    countries: []
}

const CountrySlice = createSlice({
    name: "country-slice",
    initialState,
    reducers: {
        setName: (state, action) => {
            state.name = action.payload
        },
        setCountries: (state, action) => {
            state.countries = action.payload
        },
        setCapitalWeather: (state, action) => {
            state.capitalWeather = action.payload
        },

    }
})


export const { setCapitalWeather, setCountries, setName } = CountrySlice.actions;
export default CountrySlice.reducer;