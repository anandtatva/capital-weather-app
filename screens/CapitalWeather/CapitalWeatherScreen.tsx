import { useSelector } from "react-redux"
import WeatherContainer from "./weather-container";

const CapitalWeatherScreen = () => {
    const country = useSelector(state => state.country);
    return (
        <WeatherContainer item={country.capitalWeather} index={0} />
    )
}

export default CapitalWeatherScreen;