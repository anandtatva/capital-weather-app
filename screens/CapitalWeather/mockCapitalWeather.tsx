import { CapitalWeatherType } from "../../types/capital-weather-type";

export const mockCapitalWeather: CapitalWeatherType = {
    temperature: 15,
    weather_icons: [
        'https://assets.weatherstack.com/images/wsymbols01_png_64/wsymbol_0025_light_rain_showers_night.png'
    ],
    wind_speed: 8,
    precip: 0.2
}