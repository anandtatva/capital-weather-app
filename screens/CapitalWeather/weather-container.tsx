import { View, Image, Text, StyleSheet, FlatList } from "react-native";
import { colors } from "../../utils";
import { CapitalWeatherType } from "../../types/capital-weather-type";

interface WeatherContainerProps {
    index: number;
    item: CapitalWeatherType;
}
const WeatherContainer = ({ index, item }: WeatherContainerProps) => {

    return (
        <View style={styles.container}>
            <View style={{ flex: 1, padding: 5 }}>
                <Image style={styles.image} source={{ uri: item.weather_icons[0] }} />
            </View>
            <View style={styles.textContainer}>
                <Text style={styles.text} testID={`Temperature-${index}`}>Temperature: {item.temperature}</Text>
                <Text style={styles.text} testID={`Precip-${index}`}>Precip: {item.precip} </Text>
                <Text style={styles.text} testID={`Wind-Speed-${index}`}>Wind Speed: {item.wind_speed}</Text>
            </View>
        </View>
    )
}

export default WeatherContainer;

const styles = StyleSheet.create({
    button: {
        backgroundColor: colors.secondary,
        marginVertical: 5
    },
    btnText: {
        color: colors.primary
    },
    container: {
        flexDirection: "row",
        padding: 10,
        backgroundColor: colors.primary
    },
    image: {
        width: 25,
        height: 25,
        borderRadius: 50
    },
    text: {
        color: colors.secondary,
        lineHeight: 20
    },
    textContainer: {
        flex: 4

    }
})