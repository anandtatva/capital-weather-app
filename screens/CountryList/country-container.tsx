import { View, Image, Text, StyleSheet, FlatList } from "react-native";
import { useSelector, useDispatch } from "react-redux"
import { CountryType } from "../../types/country-type";
import Button from "../../components/Button";
import { colors } from "../../utils";

interface CountryContainerProps {
    index: number;
    item: CountryType;
    onPress: (item: CountryType) => void;
}
const CountryContainer = ({ index, item, onPress }: CountryContainerProps) => {
    const handleWeather = async () => {
        await onPress(item)
    }

    return (
        <View style={styles.container}>
            <View style={{ flex: 1, padding: 5 }}>
                <Image style={styles.image} source={{ uri: item.flag }} />
            </View>
            <View style={styles.textContainer}>
                <Text style={styles.text} testID={`Capital-${index}`}>Capital: {item.capital}</Text>
                <Text style={styles.text} testID={`Population-${index}`}>Population: {item.population} </Text>
                <Text style={styles.text} testID={`LatLng-${index}`}>LatLng: {item.latlng.join(",")}</Text>
                <Button
                    rootStyle={styles.button}
                    btnTextStyle={styles.btnText}
                    onButtonPress={handleWeather}
                    name={"Captal Weather"}
                    testId={`Captal-Weather-Button-${index}`}
                />
            </View>
        </View>
    )
}

export default CountryContainer;

const styles = StyleSheet.create({
    button: {
        backgroundColor: colors.secondary,
        marginVertical: 5
    },
    btnText: {
        color: colors.primary
    },
    container: {
        flexDirection: "row",
        padding: 10,
        backgroundColor: colors.primary
    },
    image: {
        width: 25,
        height: 25,
        borderRadius: 50
    },
    text: {
        color: colors.secondary,
        lineHeight: 20
    },
    textContainer: {
        flex: 4

    }
})