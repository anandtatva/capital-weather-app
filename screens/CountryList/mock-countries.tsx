import { CountryType } from "../../types/country-type";

export const mockContries: CountryType[] = [
    {
        capital: 'Diego Garcia',
        latlng: [-6, 71.5],
        flag: 'https://flagcdn.com/io.svg',
        population: 3000
    },
    {
        capital: 'New Delhi',
        latlng: [20, 77],
        flag: 'https://flagcdn.com/in.svg',
        population: 1380004385
    }
]