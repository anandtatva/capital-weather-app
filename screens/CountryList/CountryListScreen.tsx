import { View, Text, StyleSheet, FlatList } from "react-native";
import { useSelector, useDispatch } from "react-redux"
import { CountryType } from "../../types/country-type";
import CountryContainer from "./country-container";
import { colors, showAlert } from "../../utils";
import { capitalWeatherAPIs } from "../../services/api";
import { setCapitalWeather } from "../../redux/reducer/country-slice";

const CountryListScreen = ({ navigation }) => {
    const country = useSelector(state => state.country);
    const dispatch = useDispatch();

    const handleCapitalWeather = async (item: CountryType) => {
        const resp = await capitalWeatherAPIs(item.capital);
        if (resp.status) {
            dispatch(setCapitalWeather(resp.data))
            navigation.navigate("CapitalWeatherScreen")
        } else showAlert(resp.message)
    }
    const renderItem = ({ index, item }: { item: CountryType; index: number; }) => (
        <CountryContainer index={index} item={item} onPress={handleCapitalWeather} />
    )
    return (
        <View>
            <FlatList
                data={country?.countries || []}
                renderItem={renderItem}
                keyExtractor={(_, index) => String(index)}
                ItemSeparatorComponent={() => <View style={styles.separator} />}
            />
        </View>
    )
}

export default CountryListScreen;

const styles = StyleSheet.create({
    container: {

    },
    separator: {
        borderBottomWidth: 1,
        borderBottomColor: colors.secondary
    }
})