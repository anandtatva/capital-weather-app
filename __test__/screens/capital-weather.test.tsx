import { render, waitFor, cleanup, fireEvent, act } from "@testing-library/react-native";
import { Provider } from "react-redux";
import { store } from "../../redux/configureStore";
import CapitalWeatherScreen from "../../screens/CapitalWeather/CapitalWeatherScreen";
import { setName, setCountries, setCapitalWeather } from "../../redux/reducer/country-slice";
import { Alert } from "react-native";
import { mockContries } from "../../screens/CountryList/mock-countries";
import { mockCapitalWeather } from "../../screens/CapitalWeather/mockCapitalWeather";

jest.mock("@react-navigation/native", () => {
    const actualNav = jest.requireActual("@react-navigation/native")
    return {
        ...actualNav,
        useFocusEffect: () => jest.fn(),
        useNavigation: () => ({
            navigate: jest.fn()
        }),
        useRoute: () => ({
            params: {}
        })
    }
})

describe("Screen: Capital Weather", () => {
    beforeEach(() => {
        fetchMock.mockIf(/^https?:\/\/example.com.*S/)
        store.dispatch(setName(""));
        store.dispatch(setCountries([]))
        store.dispatch(setCapitalWeather(null))
    })
    afterEach(() => {
        jest.clearAllMocks();
        cleanup()
    })
    it("renders correctly", () => {
        store.dispatch(setCapitalWeather(mockCapitalWeather))
        const tree = render(
            <Provider store={store}>
                <CapitalWeatherScreen />
            </Provider>
        ).toJSON()
        expect(tree).toMatchSnapshot();
    })
    it("Countries should be render correctly", () => {
        store.dispatch(setCapitalWeather(mockCapitalWeather))
        const tree = render(
            <Provider store={store}>
                <CapitalWeatherScreen />
            </Provider>
        ).toJSON()
        expect(tree).toMatchSnapshot();
    })

    it("Valid weather container details should be  show in capital weather container", () => {
        store.dispatch(setCapitalWeather(mockCapitalWeather))
        const { getByTestId } = render(
            <Provider store={store}>
                <CapitalWeatherScreen />
            </Provider>
        );
        const temperature = getByTestId("Temperature-0");
        const precip = getByTestId("Precip-0");
        const windSpeed = getByTestId("Wind-Speed-0");
        expect(temperature.children[1]).toBe(String(mockCapitalWeather.temperature))
        expect(precip.children[1]).toBe(String(mockCapitalWeather.precip))
        expect(windSpeed.children[1]).toBe(String(mockCapitalWeather.wind_speed))
    })
})