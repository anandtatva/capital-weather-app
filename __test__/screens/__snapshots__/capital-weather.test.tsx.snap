// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen: Capital Weather Countries should be render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#000",
      "flexDirection": "row",
      "padding": 10,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "padding": 5,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://assets.weatherstack.com/images/wsymbols01_png_64/wsymbol_0025_light_rain_showers_night.png",
        }
      }
      style={
        Object {
          "borderRadius": 50,
          "height": 25,
          "width": 25,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "flex": 4,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFF",
          "lineHeight": 20,
        }
      }
      testID="Temperature-0"
    >
      Temperature: 
      15
    </Text>
    <Text
      style={
        Object {
          "color": "#FFF",
          "lineHeight": 20,
        }
      }
      testID="Precip-0"
    >
      Precip: 
      0.2
       
    </Text>
    <Text
      style={
        Object {
          "color": "#FFF",
          "lineHeight": 20,
        }
      }
      testID="Wind-Speed-0"
    >
      Wind Speed: 
      8
    </Text>
  </View>
</View>
`;

exports[`Screen: Capital Weather renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#000",
      "flexDirection": "row",
      "padding": 10,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "padding": 5,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://assets.weatherstack.com/images/wsymbols01_png_64/wsymbol_0025_light_rain_showers_night.png",
        }
      }
      style={
        Object {
          "borderRadius": 50,
          "height": 25,
          "width": 25,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "flex": 4,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFF",
          "lineHeight": 20,
        }
      }
      testID="Temperature-0"
    >
      Temperature: 
      15
    </Text>
    <Text
      style={
        Object {
          "color": "#FFF",
          "lineHeight": 20,
        }
      }
      testID="Precip-0"
    >
      Precip: 
      0.2
       
    </Text>
    <Text
      style={
        Object {
          "color": "#FFF",
          "lineHeight": 20,
        }
      }
      testID="Wind-Speed-0"
    >
      Wind Speed: 
      8
    </Text>
  </View>
</View>
`;
